streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -130,7 +132,12 @@ public String topic() {
         if (task.record() == null)
             throw new IllegalStateException("This should not happen as topic() should only be called while a record is processed");
 
-        return task.record().topic();
+        String topic = task.record().topic();
```

Comment:
This fixes the behavior of topic, but the behavior of timestamp still looks incorrect. Maybe adding a unit test to check all the methods during a `punctuate()` call would help ensure we have the correct behavior?